### ISSUE TITLE
Range operator and magic

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9310]
Migrated from [rt.perl.org#53554](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=53554) (status was 'resolved')

Searchable as RT53554$

**Comment by p5pRT** [comment]
### From p5p@perl.wizbit.be

Double/Triple magic happens with the range operator\.

$ cat range\.pl
\#\!/usr/bin/perl \-l

use strict;
use warnings;

use Tie&#8203;::Scalar;
use base qw/Tie&#8203;::StdScalar/;
my $count;
sub FETCH \{ $count\+\+; $\{ $\_\[0\] \} \}

my $x;
tie $x\, "main"\, 5;

$count = 0;
my @&#8203;y1 = $x \.\. 6;
print "my \\@&#8203;y1 = \\$x \.\.  6 = $count calls to FETCH";

$count = 0;
my @&#8203;y2 = 4 \.\. $x;
print "my \\@&#8203;y2 =  4 \.\. \\$x = $count calls to FETCH";

$count = 0;
my @&#8203;y3 = $x \.\. $x;
print "my \\@&#8203;y3 = \\$x \.\. \\$x = $count calls to FETCH";


$ perl range\.pl
my @&#8203;y1 = $x \.\.  6 = 3 calls to FETCH
my @&#8203;y2 =  4 \.\. $x = 3 calls to FETCH
my @&#8203;y3 = $x \.\. $x = 6 calls to FETCH

\(Tested with 5\.8\.8/5\.10\.0/blead\)

I did try looking at it a while back but I gave up\.\.\.


Kind regards\,

Bram

**Comment by p5pRT** [comment]
### From @davidnicol

On Wed\, Apr 30\, 2008 at 4&#8203;:35 AM\, via RT Bram \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

>  my @&#8203;y1 = $x \.\.  6 = 3 calls to FETCH

1&#8203;: check to see if $x is an integer or a string
2&#8203;: check to see if $x is in fact less than 6
3&#8203;: read initial value

what's the problem?

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @davidnicol

On Wed\, Apr 30\, 2008 at 4&#8203;:05 PM\, David Nicol \<davidnicol@&#8203;gmail\.com> wrote&#8203;:
> On Wed\, Apr 30\, 2008 at 4&#8203;:35 AM\, via RT Bram \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>
>  >  my @&#8203;y1 = $x \.\.  6 = 3 calls to FETCH
>
>  1&#8203;: check to see if $x is an integer or a string
>  2&#8203;: check to see if $x is in fact less than 6
>  3&#8203;: read initial value

it is going to\, on the happy path\, have to keep a local variable\,
which gets loaded in
fetch \#3\, so if that was made fetch\#1 the other two checks could go
against the local
copy\.  Is range iteration against a changing target supported in an
array context? I mean
something like

&nbsp;    for my $x \( 1 \.\. $last \) \{
&nbsp;     KeepGoingLonger and $last \+= 5;
&nbsp;     \.\.\.
&nbsp;    \}

**Comment by p5pRT** [comment]
### From p5p@perl.wizbit.be

Quoting David Nicol \<davidnicol@&#8203;gmail\.com>&#8203;:

> On Wed\, Apr 30\, 2008 at 4&#8203;:05 PM\, David Nicol \<davidnicol@&#8203;gmail\.com> wrote&#8203;:
>> On Wed\, Apr 30\, 2008 at 4&#8203;:35 AM\, via RT Bram   
>> \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
>>
>>  >  my @&#8203;y1 = $x \.\.  6 = 3 calls to FETCH
>>
>>  1&#8203;: check to see if $x is an integer or a string
>>  2&#8203;: check to see if $x is in fact less than 6
>>  3&#8203;: read initial value
>
> it is going to\, on the happy path\, have to keep a local variable\,
> which gets loaded in
> fetch \#3\, so if that was made fetch\#1 the other two checks could go
> against the local
> copy\.  Is range iteration against a changing target supported in an
> array context? I mean
> something like
>
>     for my $x \( 1 \.\. $last \) \{
>      KeepGoingLonger and $last \+= 5;
>      \.\.\.
>     \}
>

for my $x \(1 \.\. $last\) \{ \}

should do the same thing as \(IMHO\)

my @&#8203;foo = 1 \.\. $last;
for my $x \(@&#8203;foo\) \{ \}

So no\, it shouldn't be supported\.

\(The same also goes for  for my $x \(reverse 1 \.\. $last\)\)


\(A patch which adds \(TODO\) tests should be coming soon\.\.\.\)

Kind regards\,

Bram

**Comment by p5pRT** [comment]
### From @nwc10

On Mon\, May 05\, 2008 at 09&#8203;:37&#8203;:10AM \+0200\, Bram wrote&#8203;:
> Quoting David Nicol \<davidnicol@&#8203;gmail\.com>&#8203;:
> 
> >On Wed\, Apr 30\, 2008 at 4&#8203;:05 PM\, David Nicol \<davidnicol@&#8203;gmail\.com> wrote&#8203;:
> >>On Wed\, Apr 30\, 2008 at 4&#8203;:35 AM\, via RT Bram   
> >>\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> >>
> >> >  my @&#8203;y1 = $x \.\.  6 = 3 calls to FETCH
> >>
> >> 1&#8203;: check to see if $x is an integer or a string
> >> 2&#8203;: check to see if $x is in fact less than 6
> >> 3&#8203;: read initial value
> >
> >it is going to\, on the happy path\, have to keep a local variable\,
> >which gets loaded in
> >fetch \#3\, so if that was made fetch\#1 the other two checks could go
> >against the local
> >copy\.  Is range iteration against a changing target supported in an
> >array context? I mean
> >something like
> >
> >    for my $x \( 1 \.\. $last \) \{
> >     KeepGoingLonger and $last \+= 5;
> >     \.\.\.
> >    \}
> >

That seems much better written out explicitly as

for \(my $x = 1; \+\+$x; $x \<= $last\) \{
&nbsp;    KeepGoingLonger and $last \+= 5;
&nbsp;    \.\.\.
\}

> for my $x \(1 \.\. $last\) \{ \}
> 
> should do the same thing as \(IMHO\)
> 
> my @&#8203;foo = 1 \.\. $last;
> for my $x \(@&#8203;foo\) \{ \}
> 
> So no\, it shouldn't be supported\.
> 
> \(The same also goes for  for my $x \(reverse 1 \.\. $last\)\)

"worksforme" \(ie yes\, I agree\. Unless and until someone pulls a really good
counter argument out of the hat\, but I don't forsee that\)

> \(A patch which adds \(TODO\) tests should be coming soon\.\.\.\)

Great\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From p5p@perl.wizbit.be

Patch with TODO test attached\.

Kind regards\,

Bram

**Comment by p5pRT** [comment]
### From p5p@perl.wizbit.be

<details><summary>Inline Patch</summary>

```diff
--- old/t/op/range.t	2008-05-05 16:33:51.000000000 +0200
+++ new/t/op/range.t	2008-05-05 19:59:56.000000000 +0200
@@ -9,7 +9,7 @@
 
 use Config;
 
-plan (115);
+plan (135);
 
 is(join(':',1..5), '1:2:3:4:5');
 
@@ -341,4 +341,66 @@
     ok($@, 'Lower bound rejected: ' . -$ii);
 }
 
+# double/tripple magic tests
+sub TIESCALAR { bless { value => $_[1], orig => $_[1] } }
+sub STORE { $_[0]{store}++; $_[0]{value} = $_[1] }
+sub FETCH { $_[0]{fetch}++; $_[0]{value} }
+sub stores { tied($_[0])->{value} = tied($_[0])->{orig};
+             delete(tied($_[0])->{store}) || 0 }
+sub fetches { delete(tied($_[0])->{fetch}) || 0 }
+    
+tie $x, "main", 6;
+
+my @foo;
+@foo = 4 .. $x;
+is(scalar @foo, 3);
+is("@foo", "4 5 6");
+{
+  local $TODO = "test for double magic with range operator";
+  is(fetches($x), 1);
+}
+is(stores($x), 0);
+
+@foo = $x .. 8;
+is(scalar @foo, 3);
+is("@foo", "6 7 8");
+{
+  local $TODO = "test for double magic with range operator";
+  is(fetches($x), 1);
+}
+is(stores($x), 0);
+
+@foo = $x .. $x + 1;
+is(scalar @foo, 2);
+is("@foo", "6 7");
+{
+  local $TODO = "test for double magic with range operator";
+  is(fetches($x), 2);
+}
+is(stores($x), 0);
+
+@foo = ();
+for (4 .. $x) {
+  push @foo, $_;
+}
+is(scalar @foo, 3);
+is("@foo", "4 5 6");
+{
+  local $TODO = "test for double magic with range operator";
+  is(fetches($x), 1);
+}
+is(stores($x), 0);
+
+@foo = ();
+for (reverse 4 .. $x) {
+  push @foo, $_;
+}
+is(scalar @foo, 3);
+is("@foo", "6 5 4");
+{
+  local $TODO = "test for double magic with range operator";
+  is(fetches($x), 1);
+}
+is(stores($x), 0);
+
 # EOF
```
</details>

**Comment by p5pRT** [comment]
### From krahnj@telus.net

Nicholas Clark wrote&#8203;:
> On Mon\, May 05\, 2008 at 09&#8203;:37&#8203;:10AM \+0200\, Bram wrote&#8203;:
>> Quoting David Nicol \<davidnicol@&#8203;gmail\.com>&#8203;:
>>>
>>> it is going to\, on the happy path\, have to keep a local variable\,
>>> which gets loaded in
>>> fetch \#3\, so if that was made fetch\#1 the other two checks could go
>>> against the local
>>> copy\.  Is range iteration against a changing target supported in an
>>> array context? I mean
>>> something like
>>>
>>>    for my $x \( 1 \.\. $last \) \{
>>>     KeepGoingLonger and $last \+= 5;
>>>     \.\.\.
>>>    \}
> 
> That seems much better written out explicitly as
> 
> for \(my $x = 1; \+\+$x; $x \<= $last\) \{
>     KeepGoingLonger and $last \+= 5;
>     \.\.\.
> \}

In your C style for loop "$x \<= $last" is a no\-op\, perhaps you meant&#8203;:

for \(my $x = 1; $x \<= $last; \+\+$x\) \{
&nbsp;     KeepGoingLonger and $last \+= 5;
&nbsp;     \.\.\.
\}



John
\-\- 
Perl isn't a toolbox\, but a small machine shop where you
can special\-order certain sorts of tools at low cost and
in short order\.                            \-\- Larry Wall

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, May 07\, 2008 at 12&#8203;:23&#8203;:41AM \-0700\, John W\. Krahn wrote&#8203;:
> Nicholas Clark wrote&#8203;:
> >On Mon\, May 05\, 2008 at 09&#8203;:37&#8203;:10AM \+0200\, Bram wrote&#8203;:
> >>Quoting David Nicol \<davidnicol@&#8203;gmail\.com>&#8203;:

> >>>   for my $x \( 1 \.\. $last \) \{
> >>>    KeepGoingLonger and $last \+= 5;
> >>>    \.\.\.
> >>>   \}
> >
> >That seems much better written out explicitly as
> >
> >for \(my $x = 1; \+\+$x; $x \<= $last\) \{
> >    KeepGoingLonger and $last \+= 5;
> >    \.\.\.
> >\}
> 
> In your C style for loop "$x \<= $last" is a no\-op\, perhaps you meant&#8203;:
> 
> for \(my $x = 1; $x \<= $last; \+\+$x\) \{
>     KeepGoingLonger and $last \+= 5;
>     \.\.\.
> \}

Yes\, I did\. \[I didn't test the code\, obviously :\-\)\]

Thanks for correcting this\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @rgs

2008/5/5 Bram \<p5p@&#8203;perl\.wizbit\.be>&#8203;:
> Patch with TODO test attached\.
>

Thanks\, applied\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'resolved' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed in f52e41ad9\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
